files/common/usr/bin/illumos-cleanup

```diff
@@ -0,0 +1,185 @@
+#!/bin/bash
+
+POST_MIGRATION_CLEANUP=/var/delphix/migration/post-migration-cleanup-required
+FINAL_ILLUMOS_CLEANUP=/var/delphix/migration/final-illumos-cleanup-required
+
+function die() {
+	echo "$(basename "$0"): $*" >&2
+	exit 1
+}
+
+function echo_err() {
+	echo "$(basename "$0"): $*" >&2
+}
+
+function usage() {
+	echo "$(basename "$0"): $*" >&2
+	echo "Usage: $(basename "$0") post-migration"
+	echo "       $(basename "$0") final"
+	exit 2
+}
+
+function destroy_dataset() {
+	local dataset="$1"
+
+	if zfs list "$dataset" &>/dev/null; then
+		if zfs destroy "$dataset"; then
+			echo "Destroyed '$dataset'"
+		else
+			echo_err "Failed to destroy '$dataset'"
+			return 1
+		fi
+	fi
+}
+
+function destroy_dataset_recursive() {
+	local dataset="$1"
+
+	if zfs list "$dataset" &>/dev/null; then
+		if zfs destroy -r "$dataset"; then
+			echo "Destroyed '$dataset' recursively"
+		else
+			echo_err "Failed to destroy '$dataset' recursively"
+			return 1
+		fi
+	fi
+}
+
+function promote_dataset() {
+	local dataset="$1"
+	local origin
+
+	#
+	# To promote a dataset all the way to the top, zfs promote may need
+	# to be called multiple times if the dataset is a nested clone.
+	# In practice though, since we delete old os versions, the clone
+	# depth shouldn't go deeper than 2 or 3 levels.
```

Comment:
why is the limit 2 or 3? what is this derived from?

Reply:
That's derived from the way we lay-out our file systems and delete older versions of upgrade images in `rpool/versions`.

For an initial Illumos install, for `/var/delphix` we have: 
`rpool/versions/[os-version]/dlpx/[dlpx-version]/install/var` that is installed, and then cloned into `rpool/versions/[os-version]/dlpx/[dlpx-version]/running/var`.

During upgrade we will have `rpool/versions/[os-version]/dlpx/[old-dlpx-version]/running/var` cloned into `rpool/versions/[os-version]/dlpx/[new-dlpx-version]/running/var`.

Then during migration we will clone `rpool/versions/[os-version]/dlpx/[dlpx-version]/running/var` into the Linux `/var/delphix` dataset.

In my testing, migrating from a fresh Delphix engine, /var/delphix required 2 promotes to get to the top. I'd expect it to require one more if we come from an already upgraded system.